bench/irmin-pack/tree.ml

```diff
@@ -405,43 +416,112 @@ struct
         config.commit_data_file
     in
     let config = { config with ncommits_trace = n } in
-    let conf = Irmin_pack.config ~readonly:false ~fresh:true config.root in
-    let* repo = Store.Repo.v conf in
+    let* repo, on_commit, repo_pp = Store.create_repo n config in
     let* result =
-      Trees_trace.add_commits repo commits |> Benchmark.run config
+      Trees_trace.add_commits repo commits on_commit |> Benchmark.run config
     in
     let+ () = Store.Repo.close repo in
+
     prepare_results_dir config.results_dir;
-    let json_channel =
+    let json_path =
       let ( / ) = Filename.concat in
-      config.results_dir / "boostrap_trace.json" |> open_out
+      config.results_dir / "boostrap_trace_timings.json"
     in
+    let json_channel = open_out json_path in
     Format.fprintf
       (Format.formatter_of_out_channel json_channel)
       "%a%!" Trees_trace.pp_stats
-      (true, config.flatten, config.inode_config);
+      (true, config.flatten, config.inode_config, config.store_type);
     close_out json_channel;
 
     fun ppf ->
       Format.fprintf ppf
-        "Tezos_log mode on inode config %a: %d commits @\n%a@\n%a"
-        pp_inode_config config.inode_config config.ncommits_trace
-        Trees_trace.pp_stats
-        (false, config.flatten, config.inode_config)
-        Benchmark.pp_results result
+        "Tezos_log mode on inode config %a, %a. @\n\
+         %a@\n\
+         Results: @\n\
+         %a@\n\
+         Stats saved to %s@\n\
+         %a"
+        pp_inode_config config.inode_config pp_store_type config.store_type
+        repo_pp () Trees_trace.pp_stats
+        (false, config.flatten, config.inode_config, config.store_type)
+        json_path Benchmark.pp_results result
 end
 
-module Bench_inodes_32 = Bench_suite (Conf)
+module Make_store_layered (Conf : sig
+  val entries : int
+  val stable_hash : int
+end) =
+struct
+  module Store =
+    Irmin_pack_layered.Make (Conf) (Irmin.Metadata.None) (Irmin.Contents.String)
+      (Irmin.Path.String_list)
+      (Irmin.Branch.String)
+      (Hash)
+
+  let create_repo ncommits config =
+    let conf = Irmin_pack.config ~readonly:false ~fresh:true config.root in
+    let* repo = Store.Repo.v conf in
+    let on_commit i commit_hash =
+      let* () =
+        if i = config.freeze_commit then
+          let* c = Store.Commit.of_hash repo commit_hash in
+          let c = Option.get c in
+          Store.freeze repo ~max:[ c ] ~min_upper:[ c ]
+        else Lwt.return_unit
+      in
+      let* () =
+        if i = ncommits - 1 then Store.PrivateLayer.wait_for_freeze repo
+        else Lwt.return_unit
+      in
+      Lwt.pause ()
+      (* Lwt_unix.sleep 0.001 *)
```

Comment:
you forgot to remove this comments;
if you think benchmarking this type of behaviour for the main thread is pertinent maybe we should add it explicitly? I'm on a fence about adding pause in the main thread, as in Tezos, we won't do this. But maybe it's useful to detect blocking behaviour in the freeze thread?

Reply:
As I mentioned in Wednesday's meeting, I highly doubt we will be able to leave the main thread's yields to chance while keeping an `Lwt.async` freeze. If we want a robust behaviour, we will at least have to change something in the way freeze works. Detecting blocking behaviours in the freeze thread would be a great start.

I don't think this parameter is pertinent yet, it will depend on the solutions we think about.

I indeed forgot to remove these comments, I will replace these with a small explanation.